README.md
# k8soauth2-proxy-controller - Proxy for dynamically swapping OAUTH2 redirect URI

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/5738/badge)](https://bestpractices.coreinfrastructure.org/projects/5738)
[![e2e](https://github.com/DoodleScheduling/k8soauth2-proxy-controller/workflows/e2e/badge.svg)](https://github.com/DoodleScheduling/k8soauth2-proxy-controller/actions)
[![report](https://goreportcard.com/badge/github.com/DoodleScheduling/k8soauth2-proxy-controller)](https://goreportcard.com/report/github.com/DoodleScheduling/k8soauth2-proxy-controller)
[![license](https://img.shields.io/github/license/DoodleScheduling/k8soauth2-proxy-controller.svg)](https://github.com/DoodleScheduling/k8soauth2-proxy-controller/blob/master/LICENSE)
[![release](https://img.shields.io/github/release/DoodleScheduling/k8soauth2-proxy-controller/all.svg)](https://github.com/DoodleScheduling/k8soauth2-proxy-controller/releases)

OAUTH2 Proxy server with kubernetes support.
The proxy is used as MitM between your idp and an external idp. The proxy dynamically replaces the redirect_uri.
This is useful if you have multiple environemnts and one or more external idp like google.
On the external idp only the oauth2 proxy needs to be configured as redirect_uri.
The proxy makes sure to route the oauth2 callbacks correctly back to your original idp.

💡 The proxy is not to be confused with a proxy which adds oidc authentication to your apps. It is rather a proxy do dynamically swap redirect_uri for external IdP in order to maintain a single redirect_uri there.

## Why?
OIDC IdP like google don't provide an API to dynamically configure or change OAUTH2 credentials. Meaning there
is no good way of progamatically add/remove redirect_uri or create ad hock credentials.
With the oauth2 proxy its possible to only have the proxy URL configured as redirect_uri.

The proxy transfers any OAUTH2 state from the original request as well as the redirect_uri in the [state param](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.1).
While the external IdP redirects back to the oauth2 proxy after a successful authorization the oauth2 proxy unpacks the state and transforms the request back to its original.


## Example OAUTH2Proxy

A `OAUTH2Proxy` binds to a kubernetes service.
The http proxy will route and clone incoming http requests to all OAUTH2Proxy backends matching the host.

```yaml
apiVersion: oauth2.infra.doodle.com/v1beta1
kind: OAUTH2Proxy
metadata:
  name: idp
spec:
  host: my-idp
  paths:
  - /
  redirectURI: https://oauth-proxy
  backend:
    serviceName: backend-idp
    servicePort: http
```

## Setup

The proxy should not be exposed directly to the public. Rather should traffic be routed via an ingress controller
and only paths which are used to redirect to the external idp should be routed via the oauth2 proxy.

### Helm chart

Please see [chart/k8soauth2-proxy-controller](https://github.com/DoodleScheduling/k8soauth2-proxy-controller) for the helm chart docs.

### Manifests/kustomize

Alternatively you may get the bundled manifests in each release to deploy it using kustomize or use them directly.

## Configure the controller

You may change base settings for the controller using env variables (or alternatively command line arguments).
Available env variables:

| Name  | Description | Default |
|-------|-------------| --------|
| `METRICS_ADDR` | The address of the metric endpoint binds to. | `:9556` |
| `PROBE_ADDR` | The address of the probe endpoints binds to. | `:9557` |
| `HTTP_ADDR` | The address of the http proxy. | `:8080` |
| `PROXY_READ_TIMEOUT` | Read timeout to the proxy backend. | `30s` |
| `PROXY_WRITE_TIMEOUT` | Write timeout to the proxy backend. | `30s` |
| `ENABLE_LEADER_ELECTION` | Enable leader election for controller manager. | `false` |
| `LEADER_ELECTION_NAMESPACE` | Change the leader election namespace. This is by default the same where the controller is deployed. | `` |
| `NAMESPACES` | The controller listens by default for all namespaces. This may be limited to a comma delimted list of dedicated namespaces. | `` |
| `CONCURRENT` | The number of concurrent reconcile workers.  | `2` |
| `OTEL_EXPORTER_OTLP_ENDPOINT` | The gRPC opentelemtry-collector endpoint uri | `` |

**Note:** The proxy implements opentelemetry tracing, see [further possible env](https://opentelemetry.io/docs/reference/specification/sdk-environment-variables/) variables to configure it.


CHANGELOG.md
# Changelog

## 0.1.2

**Release date:** 2022-07-07

### Features
* Support IdP which sends the redirect request via POST, transform request into GET with query strings


## 0.1.1

**Release date:** 2022-06-15

### Fixes
* Set USER env in image as otel requires it


## 0.1.0

**Release date:** 2022-06-15

Initial oss release


main.go
/*


Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"context"
	"flag"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/spf13/pflag"
	"github.com/spf13/viper"
	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/contrib/propagators/b3"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/sdk/resource"
	"go.opentelemetry.io/otel/sdk/trace"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/runtime"
	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/cache"
	"sigs.k8s.io/controller-runtime/pkg/healthz"
	"sigs.k8s.io/controller-runtime/pkg/log/zap"

	infradoodlecomv1beta1 "github.com/DoodleScheduling/k8soauth2-controller/api/v1beta1"
	"github.com/DoodleScheduling/k8soauth2-controller/controllers"
	"github.com/DoodleScheduling/k8soauth2-controller/proxy"
	// +kubebuilder:scaffold:imports
)

var (
	scheme   = runtime.NewScheme()
	setupLog = ctrl.Log.WithName("setup")
)

func init() {
	_ = clientgoscheme.AddToScheme(scheme)

	_ = corev1.AddToScheme(scheme)
	_ = infradoodlecomv1beta1.AddToScheme(scheme)
	// +kubebuilder:scaffold:scheme
}

var (
	proxyReadTimeout        = "30s"
	proxyWriteTimeout       = "30s"
	httpAddr                = ":8080"
	metricsAddr             = ":9556"
	probesAddr              = ":9557"
	enableLeaderElection    = false
	leaderElectionNamespace string
	namespaces              = ""
	concurrent              = 2
)

func main() {
	flag.StringVar(&metricsAddr, "metrics-addr", ":9556", "The address of the metric endpoint binds to.")
	flag.StringVar(&probesAddr, "probe-addr", ":9557", "The address of the probe endpoints bind to.")
	flag.StringVar(&httpAddr, "http-addr", ":8080", "The address of http server binding to.")
	flag.StringVar(&proxyReadTimeout, "proxy-read-timeout", "10s", "Read timeout for proxy requests.")
	flag.StringVar(&proxyWriteTimeout, "proxy-write-timeout", "10s", "Write timeout for proxy requests.")
	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
		"Enable leader election for controller manager. "+
			"Enabling this will ensure there is only one active controller manager.")
	flag.StringVar(&leaderElectionNamespace, "leader-election-namespace", "",
		"Specify a different leader election namespace. It will use the one where the controller is deployed by default.")
	flag.StringVar(&namespaces, "namespaces", "",
		"The controller listens by default for all namespaces. This may be limited to a comma delimted list of dedicated namespaces.")
	flag.IntVar(&concurrent, "concurrent", 2,
		"The number of concurrent reconcile workers. By default this is 2.")

	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
	pflag.Parse()
	ctrl.SetLogger(zap.New())

	// Import flags into viper and bind them to env vars
	// flags are converted to upper-case, - is replaced with _
	err := viper.BindPFlags(pflag.CommandLine)
	if err != nil {
		setupLog.Error(err, "Failed parsing command line arguments")
		os.Exit(1)
	}

	replacer := strings.NewReplacer("-", "_")
	viper.SetEnvKeyReplacer(replacer)
	viper.AutomaticEnv()

	opts := ctrl.Options{
		Scheme:                  scheme,
		MetricsBindAddress:      viper.GetString("metrics-addr"),
		HealthProbeBindAddress:  viper.GetString("probe-addr"),
		LeaderElection:          viper.GetBool("enable-leader-election"),
		LeaderElectionNamespace: viper.GetString("leader-election-namespace"),
		LeaderElectionID:        "k8soauth2-proxy-controller",
	}

	ns := strings.Split(viper.GetString("namespaces"), ",")
	if len(ns) > 0 && ns[0] != "" {
		opts.NewCache = cache.MultiNamespacedCacheBuilder(ns)
		setupLog.Info("watching dedicated namespaces", "namespaces", ns)
	} else {
		setupLog.Info("watching all namespaces")
	}

	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), opts)
	if err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}

	// Add liveness probe
	err = mgr.AddHealthzCheck("healthz", healthz.Ping)
	if err != nil {
		setupLog.Error(err, "Could not add liveness probe")
		os.Exit(1)
	}

	// Add readiness probe
	err = mgr.AddReadyzCheck("readyz", healthz.Ping)
	if err != nil {
		setupLog.Error(err, "Could not add readiness probe")
		os.Exit(1)
	}

	resources, err := resource.New(context.Background(),
		resource.WithFromEnv(), // pull attributes from OTEL_RESOURCE_ATTRIBUTES and OTEL_SERVICE_NAME environment variables
		resource.WithProcess(), // This option configures a set of Detectors that discover process information
	)
	if err != nil {
		setupLog.Error(err, "failed creating OTLP resources")
		os.Exit(1)
	}

	client := otlptracegrpc.NewClient()
	exporter, err := otlptrace.New(context.Background(), client)
	if err != nil {
		setupLog.Error(err, "failed creating OTLP trace exporter")
		os.Exit(1)
	}

	tp := trace.NewTracerProvider(
		trace.WithBatcher(exporter),
		trace.WithResource(resources),
	)

	otel.SetTextMapPropagator(b3.New())
	defer func() {
		if err := tp.Shutdown(context.Background()); err != nil {
			setupLog.Error(err, "failed to shutdown trace provider")
		}
	}()

	otel.SetTracerProvider(tp)
	proxy := proxy.New(setupLog, &http.Client{
		Transport: otelhttp.NewTransport(http.DefaultTransport),
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	})

	wrappedHandler := otelhttp.NewHandler(proxy, "oauth2-proxy")

	readTimeout, err := time.ParseDuration(viper.GetString("proxy-read-timeout"))
	if err != nil {
		setupLog.Error(err, "failed to parse proxy read timeout")
	}

	writeTimeout, err := time.ParseDuration(viper.GetString("proxy-write-timeout"))
	if err != nil {
		setupLog.Error(err, "failed to parse proxy write timeout")
	}

	s := &http.Server{
		Addr:         httpAddr,
		Handler:      wrappedHandler,
		ReadTimeout:  readTimeout,
		WriteTimeout: writeTimeout,
	}

	go s.ListenAndServe()

	pReconciler := &controllers.OAUTH2ProxyReconciler{
		Client:    mgr.GetClient(),
		Log:       ctrl.Log.WithName("controllers").WithName("OAUTH2Proxy"),
		Scheme:    mgr.GetScheme(),
		Recorder:  mgr.GetEventRecorderFor("OAUTH2Proxy"),
		HttpProxy: proxy,
	}
	if err = pReconciler.SetupWithManager(mgr, controllers.OAUTH2ProxyReconcilerOptions{MaxConcurrentReconciles: viper.GetInt("concurrent")}); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "OAUTH2Proxy")
		os.Exit(1)
	}

	// +kubebuilder:scaffold:builder

	setupLog.Info("starting manager")
	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
		setupLog.Error(err, "problem running manager")
		os.Exit(1)
	}
}
